### DISCUSSION TITLE
How should I publish with meta information and resources?

### DISCUSSION THREAD

**Discussion by oovm** [bytecodealliance/registry#280] [Q&A] [answered]
I want to know what is the positioning of the published unit (*.wasm)?

Is it similar to the jar on the jvm platform?

If I want to publish a library, where should `Manifest.xml` in jvm or `package.json` in nodejs be stored?

Is it stored directly in the `custom-section` of wasm?

If my program requires some static resources, such as images, should I directly transcode and store them in the `data-section`?

**Comment by macovedj** [comment, marked as answer]
Currently it is possible to add metadata to a binary's custom section via the `wasm-tools metadata` command, the code for which can be find in [this crate](https://github.com/bytecodealliance/wasm-tools/tree/main/crates/wasm-metadata).  There are also plans to render this information on wa.dev, as well as to make it easier to modify in the UI as well as via the warg client and local config. 

After you have time to check out the information currently supported in the metadata, I'll be curious if there is other information found in language specific package files that you had hoped would be present that isn't.  At the moment there isn't a plan to include language specific package files, though wa.dev will likely support configuration for publishing to registries associated with source languages (like crates or npm) in the future.  You may also be curious to follow along the [larger conversation](https://github.com/bytecodealliance/wasm-pkg-tools/discussions/3) happening around how configuration should work for wasm registries.

I don't have a great answer re: static resources today, but I think perhaps [WASI-Virt](https://github.com/bytecodealliance/WASI-Virt) could be helpful for that use case if not today then in the future.

**Comment by oovm** [comment]
OK, I noticed the [registry-metadata](https://github.com/bytecodealliance/wasm-tools/blob/443fc549fd34d66765a682dffa7d5f4a599b3ad6/crates/wasm-metadata/src/lib.rs#L470 ) field, and I checked it Definition, if I understand correctly, this stores meta-information about wa.dev, not language-specific meta-information.

I am currently writing a wrapping package release tool. I hope that when releasing the library, the source code in the original language can be carried, so that users can modify and recompile the library after downloading it for testing. 

Should I submit a PR to occupy a field at the same level of `registry-metadata` field?

**Reply by calvinrp** [reply]
We probably need a different solution than putting the source code in a Wasm custom section. Interesting problem, especially around reproducible builds.

**Reply by lann** [reply]
I would suggest putting the source in a _new_ custom section, perhaps as a tar+gzip bundle. It does seem generally useful to have this as an option. https://github.com/bytecodealliance/wasm-pkg-tools/issues/12
